src/models/data.py

```diff
@@ -169,43 +226,109 @@ def ds_folder_to_np(self,
         assert len(list(y.data_vars)) == 1, f'Expect only 1 target variable! ' \
             f'Got {len(list(y.data_vars))}'
+
+        coords = [c for c in y.coords]
+        target_var = [y for y in y.variables if y not in coords][0]
+
         x_np, y_np = x.to_array().values, y.to_array().values
 
-        if (self.normalizing_dict is not None) and (self.normalizing_array is None):
-            self.calculate_normalizing_array(list(x.data_vars))
+        if (self.normalizing_dict is not None) and (self.historical_normalizing_array is None):
+            self.calculate_normalizing_array(list(x.data_vars), target_var=target_var)
 
         # first, x
         x_np = x_np.reshape(x_np.shape[0], x_np.shape[1], x_np.shape[2] * x_np.shape[3])
         x_np = np.moveaxis(np.moveaxis(x_np, 0, 1), -1, 0)
+
         # then, the x month
-        assert len(y.time) == 1, f'Expected y to only have 1 timestamp! Got {len(y.time)}'
-        target_month = datetime.strptime(str(y.time.values[0])[:-3], '%Y-%m-%dT%H:%M:%S.%f').month
+        assert len(y.time) == 1, 'Expected y to only have 1 timestamp!'\
+            f'Got {len(y.time)}'
+        target_month = datetime.strptime(
+            str(y.time.values[0])[:-3], '%Y-%m-%dT%H:%M:%S.%f'
+        ).month
         x_months = np.array([target_month] * x_np.shape[0])
+
         # then, y
         y_np = y_np.reshape(y_np.shape[0], y_np.shape[1], y_np.shape[2] * y_np.shape[3])
         y_np = np.moveaxis(y_np, -1, 0).reshape(-1, 1)
 
-        if self.normalizing_array is not None:
-            x_np = (x_np - self.normalizing_array['mean']) / (self.normalizing_array['std'])
+        if self.experiment == 'nowcast':
+            # if nowcast then we have a TrainData.current
+            historical = x_np[:, :-1, :]  # all timesteps except the final
+            current = x_np[:, -1, :-1]  # final axis is all nan (target var)
+
+            # add 1 axis for timesteps
+            current = current.reshape(x_np.shape[0], 1, current.shape[-1])
+            train_data = TrainData(
+                current=current,
+                historical=historical,
+                pred_months=x_months  # @GABI do we want to include this ?
+            )
+        else:
+            train_data = TrainData(
+                current=None,
+                historical=x_np,
+                pred_months=x_months
+            )
+
+        if self.historical_normalizing_array is not None:
+            if self.experiment == 'nowcast':
+                # normalize the historical X data (all vars including target variable)
+                train_data.historical = (
+                    train_data.historical - self.historical_normalizing_array['mean']
+                ) / (self.historical_normalizing_array['std'])
+
+                # normalize the current X data (no target variable)
+                train_data.current = (
+                    train_data.current - self.current_normalizing_array['mean']
```

Comment:
The mypy errors here are because mypy still thinks `self.current_normalizing_array` could be `None` (since you don't technically check that its not in the statement above).

Mypy could be appeased if you changed the above statement to:

`if (self.historical_normalizing_array is not None) and (self.current_normalizing_array is not None):`

which should always be `True` if the current statement is

Reply:
okay i think i fixed that